### PR TITLE
ggml: add Q1_0 1-bit quantization support (CPU)

### PR DESCRIPTION
**Pull request by khosravipasha**
## Overview

**Update**:  Initially had two types of Q1_0 (group size 32) and Q1_0_g128 (group size 128) but now only keeping Q1_0 with group size 128. If you see Q1_0_g128 in screenshots anywhere its now Q1_0.

This PR adds `Q1_0` a 1-bit quantization/packing support for CPU. Motivation is to support inference for family of [1-bit Bonsai models ](https://huggingface.co/collections/prism-ml/bonsai) which are native in `Q1_0` (group size 128).

Basically each group has a shared scale of fp16, and then the weights are packed into 1-bit, i.e. each weight is represented either by a 0 or 1. The 0 would represent -scale, and the 1 would represent +scale.

We just released Bonsai 1-bit models, which are 3 models that are natively in Q1_0 format.

With help of llama.cpp can run our model at a very fast speed while being decent at evals. (see cpu benchmarks below). I have not really tried optimizing the cpu version as much, mostly tested for correctness, our main focus was Metal/CUDA which will come later in future PRs.

Speeds across devices:
Screenshot 2026-04-01 at 09 48 39

Evals comparing to simalar param count models (but at a much smaller footprint)
Screenshot 2026-04-01 at 09 48 46

### Models

The Q1_0_g128 models can be found here, in this PR mostly used the 1.7B model for testing since this is CPU only.
- [prism-ml/Bonsai-8B-gguf](https://huggingface.co/prism-ml/Bonsai-8B-gguf)
- [prism-ml/Bonsai-4B-gguf](https://huggingface.co/prism-ml/Bonsai-4B-gguf)
- [prism-ml/Bonsai-1.7B-gguf](https://huggingface.co/prism-ml/Bonsai-1.7B-gguf)

- See a working demo from our Bonsai-demo repo: [Bonsai-demo](https://github.com/prismml-eng/Bonsai-demo)
- See more details here: [1-bit-bonsai-8b-whitepaper.pdf](https://github.com/PrismML-Eng/Bonsai-demo/blob/main/1-bit-bonsai-8b-whitepaper.pdf)

### Testing Process

We test two cpu routes: ARM NEON and Generic Scalar Fallback. Everything tested on a Mac M4 Pro 48 GB.

To build for Generic Scalar Fallback we temporarily removed NEON paths and then rebuilt and ran KL divergences (packed vs unpacked), speed benchmarks, prompted the model to see if it can give coherent output.

Tried above for following combinations:
 -  Q1_0 vs old_Q1_0; (old_Q1_0 is the one with group size 32; not committing it but showing the difference in speed and size)
 -  ARM NEON vs Generic Scalar Fallback;
 
<details>
<summary>Benchmarks</summary>

ARM NEON 
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| qwen3 1.7B Q1_0            | 231.13 MiB |     1.72 B | BLAS       |      10 |           pp512 |       502.21 ± 13.70 |
| qwen3 1.7B Q1_0           | 231.13 MiB |     1.72 B | BLAS       |      10 |           tg128 |        112.46 ± 5.89 |
| qwen3 1.7B old_Q1_0                | 308.02 MiB |     1.72 B | BLAS       |      10 |           pp512 |        505.17 ± 6.27 |
| qwen3 1.7B old_Q1_0                | 308.02 MiB |     1.72 B | BLAS       |      10 |           tg128 |       105.15 ± 12.44 |

Generic Scalar Fallback Benchmark
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| qwen3 1.7B Q1_0          | 231.13 MiB |     1.72 B | BLAS       |      10 |           pp512 |        504.63 ± 8.08 |
| qwen3 1.7B Q1_0           | 231.13 MiB |     1.72 B | BLAS       |      10 |           tg128 |         32.65 ± 2.04 |
| qwen3 1.7B old_Q1_0                | 308.02 MiB |     1.72 B | BLAS       |      10 |           pp512 |        508.45 ± 5.89 |
| qwen3 1.7B old_Q1_0                | 308.02 MiB |     1.72 B | BLAS       |      10 |           tg128 |         29.80 ± 0.38 |
</details>

#### Testing correctness of new kernels end-to-ned

Download:
- [prism-ml/Bonsai-1.7B-gguf](https://huggingface.co/prism-ml/Bonsai-1.7B-gguf)
- [prism-ml/Bonsai-1.7B-unpacked](https://huggingface.co/prism-ml/Bonsai-1.7B-unpacked)

<details>
<summary>Or follow Instructions how to pack into GGUF Q1_0</summary>

Start from FP16 HF checkpoint downloaded from [prism-ml/Bonsai-1.7B-unpacked](https://huggingface.co/prism-ml/Bonsai-1.7B-unpacked), then convert to FP16 GGUF:
```bash
python convert_hf_to_gguf.py models/Bonsai-1.7B-unpacked --outfile models/Bonsai-1.7B-f16.gguf --outtype f16 --model-name "Bonsai-1.7B"
```

When packing to Q1_0_g128 (Q1_0), mainly remember to use `--pure`. Since group size for model is 128, can also pack it into 32 without problem (since 128 is divisible by 32).

```bash
./build/bin/llama-quantize --pure models/Bonsai-1.7B-f16.gguf models/Bonsai-1.7B-Q1_0_g128.gguf Q1_0_g128
# model size      =  3280.93 MiB (16.00 BPW)
# Q1_0_g128 size  =   231.13 MiB (1.125 BPW)

./build/bin/llama-quantize --pure models/Bonsai-1.7B-f16.gguf models/Bonsai-1.7B-Q1_0.gguf Q1_0
# model size  =  3280.93 MiB (16.00 BPW)
# Q1_0 size   =   308.02 MiB (1.50 BPW)
```
</details>

#### KL Divergences of packed vs unpacked 
For more end-to-end testing of prompt processing, we can use logit dump from llama-perplexity. We compare the unpacked F16 GGUF with the packed Q1_0_g128 (or Q1_0) GGUF. The weights are identical but different formats so logits should be mostly the same. Mostly tests prompt-processing though.

<details>
<summary>How to test KL divergence packed vs unpacked</summary>

```bash
# Step 1: Save F16 reference logits
./build/bin/llama-perplexity -m models/Bonsai-1.7B-f16.gguf \
    -f datasets/wikitext-2-raw/wiki.test.raw -c 512 --chunks 100 \
    --save-all-logits models/f16_logits.bin

# Step 2: KL divergence Q1_0_g128 vs F16
./build/bin/llama-perplexity -m models/Bonsai-1.7B-Q1_0_g128.gguf \
    -f datasets/wikitext-2-raw/wiki.test.raw -c 512 --chunks 100 \
    --kl-divergence --kl-divergence-base models/f16_logits.bin

# Step 3: KL divergence Q1_0 vs F16
./build/bin/llama-perplexity -m models/Bonsai-1.7B-Q1_0.gguf \
    -f datasets/wikitext-2-raw/wiki.test.raw -c 512 --chunks 100 \
    --kl-divergence --kl-divergence-base models/f16_logits.bin
```

</details>

<details>
<summary>Divergences</summary>

ARM NEON

| Metric           | Q1_0_g128 (1.13 BPW)  | Q1_0 (1.50 BPW)       |
|:-----------------|:----------------------:|:----------------------:|
| Same top p       |    99.965 ± 0.012 %    |    99.973 ± 0.010 %    |
| Mean KLD         | 0.000000 ± 0.000000    | 0.000000 ± 0.000000    |
| Maximum KLD      |            0.000065    |            0.000065    |
| 99.9% KLD        |            0.000051    |            0.000051    |
| 99.0% KLD        |            0.000036    |            0.000036    |
| Median KLD       |            0.000000    |            0.000000    |
| 1.0% KLD         |           -0.000036    |           -0.000036    |
| Minimum KLD      |           -0.000061    |           -0.000061    |
| Mean Δp          |     -0.000 ± 0.000 %   |     -0.000 ± 0.000 %   |
| Maximum Δp       |              0.120 %   |              0.122 %   |
| 99.9% Δp         |              0.039 %   |              0.038 %   |
| 99.0% Δp         |              0.017 %   |              0.017 %   |
| 95.0% Δp         |              0.007 %   |              0.008 %   |
| Median Δp        |              0.000 %   |              0.000 %   |
| 5.0% Δp          |             -0.008 %   |             -0.007 %   |
| 1.0% Δp          |             -0.017 %   |             -0.018 %   |
| 0.1% Δp          |             -0.039 %   |             -0.038 %   |
| Minimum Δp       |             -0.102 %   |             -0.122 %   |
| RMS Δp           |      0.006 ± 0.000 %   |      0.006 ± 0.000 %   |

Generic Scalar Fallback Divergences

| Metric           | Q1_0_g128 (1.13 BPW)  | Q1_0 (1.50 BPW)       |
|:-----------------|:----------------------:|:----------------------:|
| Same top p       |    99.965 ± 0.012 %    |    99.973 ± 0.010 %    |
| Mean KLD         | 0.000000 ± 0.000000    | 0.000000 ± 0.000000    |
| Maximum KLD      |            0.000065    |            0.000065    |
| 99.9% KLD        |            0.000051    |            0.000051    |
| 99.0% KLD        |            0.000036    |            0.000036    |
| Median KLD       |            0.000000    |            0.000000    |
| 1.0% KLD         |           -0.000036    |           -0.000036    |
| Minimum KLD      |           -0.000061    |           -0.000061    |
| Mean Δp          |     -0.000 ± 0.000 %   |     -0.000 ± 0.000 %   |
| Maximum Δp       |              0.120 %   |              0.122 %   |
| 99.9% Δp         |              0.039 %   |              0.038 %   |
| 99.0% Δp         |              0.017 %   |              0.017 %   |
| 95.0% Δp         |              0.007 %   |              0.008 %   |
| Median Δp        |              0.000 %   |              0.000 %   |
| 5.0% Δp          |             -0.008 %   |             -0.007 %   |
| 1.0% Δp          |             -0.017 %   |             -0.018 %   |
| 0.1% Δp          |             -0.039 %   |             -0.038 %   |
| Minimum Δp       |             -0.102 %   |             -0.122 %   |
| RMS Δp           |      0.006 ± 0.000 %   |      0.006 ± 0.000 %   |
</details>

# Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md): Yes

- AI usage disclosure: Yes, NEON paths were generated with help of AI. The generic paths were manual by following Q4_0 as a template. The NEON paths were throughly tested as seen above and are correct (and is 4-5x faster in token generation than generic fallback, I did not test on x86 so no AVX kernels included)

### CODE REVIEW

**Comment by ggerganov** [comment]
Same as @0cc4m, I am also wondering if 32-block size is warranted to be added atm, given that only 128-block size is being used. @khosravipasha Do you have any insights about future models and do you expect smaller block size to be needed?

The name `Q1_0_g128` is very inconsistent. It should be changed to something in the format `Q1_Y` where `Y` is a digit or a letter. The conventions that we have are:

- `QX_0` - single scale
- `QX_1` - scale + bias
- `QX_K` - superblocks

File: ggml/src/ggml-common.h [hunk 1]
```diff
@@ -170,6 +174,13 @@ typedef sycl::half2 ggml_half2;
 #define GGML_EXTENSION __extension__
 #endif // _MSC_VER
 
+#define QK1_0 128
+typedef struct {
+    ggml_half d;               // delta
+    uint8_t qs[QK1_0 / 8]; // bits / quants
```
Comment (CISC) [inline ggml/src/ggml-common.h]:
```suggestion
    ggml_half d;           // delta
    uint8_t qs[QK1_0 / 8]; // bits / quants
```

File: ggml/src/ggml.c [hunk 2]
```diff
@@ -7652,6 +7661,7 @@ size_t ggml_quantize_chunk(
     size_t result = 0;
 
     switch (type) {
+        case GGML_TYPE_Q1_0: result = quantize_q1_0(src + start, (char *) dst + start_row * row_size, nrows, n_per_row, imatrix); break;
         case GGML_TYPE_Q4_0:    result = quantize_q4_0(src + start, (char *) dst + start_row * row_size, nrows, n_per_row, imatrix); break;
```
Comment (CISC) [inline ggml/src/ggml.c]:
```suggestion
        case GGML_TYPE_Q1_0:    result = quantize_q1_0(src + start, (char *) dst + start_row * row_size, nrows, n_per_row, imatrix); break;
        case GGML_TYPE_Q4_0:    result = quantize_q4_0(src + start, (char *) dst + start_row * row_size, nrows, n_per_row, imatrix); break;
```

File: gguf-py/gguf/constants.py [hunk 3]
```diff
@@ -3965,6 +3966,7 @@ class LlamaFileType(IntEnum):
     MOSTLY_TQ2_0         = 37  # except 1d tensors
     MOSTLY_MXFP4_MOE     = 38  # except 1d tensors
     MOSTLY_NVFP4         = 39  # except 1d tensors
+    MOSTLY_Q1_0          = 41  # except 1d tensors
```
Comment (CISC) [inline gguf-py/gguf/constants.py]:
```suggestion
    MOSTLY_Q1_0          = 40  # except 1d tensors
```

File: gguf-py/gguf/constants.py [hunk 4]
```diff
@@ -4074,6 +4076,7 @@ class VisionProjectorType:
     GGMLQuantizationType.TQ2_0:   (256, 2 + 64),
     GGMLQuantizationType.MXFP4:   (32, 1 + 16),
     GGMLQuantizationType.NVFP4:   (64, 4 + 32),
+    GGMLQuantizationType.Q1_0: (128, 2 + 16),  # 2 bytes fp16 scale + 16 bytes (128 bits)
```
Comment (CISC) [inline gguf-py/gguf/constants.py]:
```suggestion
    GGMLQuantizationType.NVFP4:   (64, 4 + 32),
    GGMLQuantizationType.Q1_0:    (128, 2 + 16),
```

File: include/llama.h [hunk 5]
```diff
@@ -154,6 +154,7 @@ extern "C" {
         LLAMA_FTYPE_MOSTLY_TQ2_0         = 37, // except 1d tensors
         LLAMA_FTYPE_MOSTLY_MXFP4_MOE     = 38, // except 1d tensors
         LLAMA_FTYPE_MOSTLY_NVFP4         = 39, // except 1d tensors
+        LLAMA_FTYPE_MOSTLY_Q1_0          = 41, // except 1d tensors
```
Comment (CISC) [inline include/llama.h]:
```suggestion
        LLAMA_FTYPE_MOSTLY_Q1_0          = 40, // except 1d tensors
```

**Review by CISC — APPROVED** [review]
Let's not leave a hole in the ftypes, changing it will just log a warning for the one GGUF already made, so no big deal.

**Comment by khosravipasha** [comment]
@CISC Thanks make sense, applied the suggested changes.
There was a CI failing for [ubuntu-24-webgpu-wasm](https://github.com/ggml-org/llama.cpp/actions/runs/23963795262/job/69899771154?pr=21273#logs). I added the generic fallback in most places was not sure where to do for this one (or might have been an older run)

**Comment by CISC** [comment]
> @CISC Thanks make sense, applied the suggested changes. There was a CI failing for [ubuntu-24-webgpu-wasm](https://github.com/ggml-org/llama.cpp/actions/runs/23963795262/job/69899771154?pr=21273#logs). I added the generic fallback in most places was not sure where to do for this one (or might have been an older run)

You're missing the definitions in `arch-fallback.h`. :)

Seems GitHub UI is up to its newline shenanigans again, it has committed the suggestions with \r\n, please clean them up:
https://github.com/ggml-org/llama.cpp/actions/runs/23997646793/job/69989182477?pr=21273

**Review by ggerganov — APPROVED** [review]

**Comment by Green-Sky** [comment]
> That's wildly impressive, though is it actually usable?

Probalby not, but q1_0 is also the lower bound of what is currently possible.

Will test some more later, but for now I ran hellaswag(400) as a proxy on the l2 7b.

| k    |   v  | r | score    | 95% confidence interval |
|------|------|---|----------|----------------------|
| f16  | f16  | 0 | 77.0000% | [72.6307%, 80.8556%] |
| q8_0 | q8_0 | 1 | 76.7500% | [72.3677%, 80.6234%] |
| q4_0 | q4_0 | 0 | 76.0000% | [71.5800%, 79.9254%] |
| q4_0 | q4_0 | 1 | 76.7500% | [72.3677%, 80.6234%] |
| q4_0 | q1_0 | 0 | 29.0000% | [24.7697%, 33.6299%] |
| q4_0 | q1_0 | 1 | 65.2500% | [60.4585%, 69.7514%] |


Anecdotally, I previously found q4_0 (both) too harsh on qwen next coder, before the rotation stuff. Which is why I included q4_0 results with and without rotation.

**Comment by ddh0** [comment]
Should tensors being quantized to `GGML_TYPE_Q1_0` require importance matrix data? I would assume so, it being such a low-bit type, but none of the other static quantization types require imatrix.
